src/main/java/uk/gov/pay/ledger/transaction/service/TransactionMetadataService.java

```diff
@@ -0,0 +1,45 @@
+package uk.gov.pay.ledger.transaction.service;
+
+import com.google.inject.Inject;
+import uk.gov.pay.ledger.event.model.EventDigest;
+import uk.gov.pay.ledger.metadatakey.dao.MetadataKeyDao;
+import uk.gov.pay.ledger.transaction.dao.TransactionDao;
+import uk.gov.pay.ledger.transactionmetadata.dao.TransactionMetadataDao;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+public class TransactionMetadataService {
+
+    private final TransactionDao transactionDao;
+    private final MetadataKeyDao metadataKeyDao;
+    private final TransactionMetadataDao transactionMetadataDao;
+
+    @Inject
+    public TransactionMetadataService(MetadataKeyDao metadataKeyDao,
+                                      TransactionMetadataDao transactionMetadataDao,
+                                      TransactionDao transactionDao) {
+        this.metadataKeyDao = metadataKeyDao;
+        this.transactionMetadataDao = transactionMetadataDao;
+        this.transactionDao = transactionDao;
+    }
+
+    public void upsertMetadataFor(EventDigest eventDigest) {
+        transactionDao.findTransactionByExternalId(eventDigest.getResourceExternalId())
+                .ifPresent(transactionEntity -> {
+                    List<String> metadataKeys = metadataKeyListFrom(eventDigest);
+                    metadataKeys.forEach(metadataKey -> {
+                        metadataKeyDao.insertIfNotExist(metadataKey);
+                        transactionMetadataDao
+                                .insertIfNotExist(transactionEntity.getId(), metadataKey);
+                    });
+                });
+    }
+
+    private List<String> metadataKeyListFrom(EventDigest eventDigest) {
+        Map<String, Object> metadata =
+                (Map<String, Object>) eventDigest.getEventPayload().get("external_metadata");
```

Comment:
Not sure if nested object is returned as Map<String,Object> too with https://github.com/alphagov/pay-ledger/blob/master/src/main/java/uk/gov/pay/ledger/event/model/EventDigest.java#L83.  Would be good to have a test if not exists and if possible.